bridge/setu/broadcaster/broadcaster_test.go

```diff
@@ -1,61 +1,296 @@
+	testCases := []struct {
+		name       string
+		msg        sdk.Msg
+		op         func(*app.HeimdallApp) error
+		expResCode uint32
+		expErr     bool
+		tearDown   func(*app.HeimdallApp) error
+	}{
+		{
+			name: "successful broadcast",
+			msg:  msgs[0],
+
+			op:         nil,
+			expResCode: 0,
+			expErr:     false,
+		},
+		{
+			name: "failed broadcast (insufficient funds for fees)",
+			msg:  msgs[1],
+			op: func(hApp *app.HeimdallApp) error {
+				acc := hApp.AccountKeeper.GetAccount(sdkCtx, hmTypes.BytesToHeimdallAddress([]byte(address)))
+				// reduce account balance to 0
+				if err := acc.SetCoins(sdk.Coins{}); err != nil {
+					return err
+				}
+				hApp.AccountKeeper.SetAccount(sdkCtx, acc)
+				return nil
+			},
+			expResCode: 5,
+			expErr:     true,
+			tearDown: func(hApp *app.HeimdallApp) error {
+				acc := hApp.AccountKeeper.GetAccount(sdkCtx, hmTypes.BytesToHeimdallAddress([]byte(address)))
+				// reset account balance
+				if err := acc.SetCoins(sdk.Coins{sdk.Coin{Denom: authTypes.FeeToken, Amount: defaultBalance}}); err != nil {
+					return err
+				}
+				hApp.AccountKeeper.SetAccount(sdkCtx, acc)
+				return nil
+			},
+		},
+		{
+			name: "failed broadcast (invalid sequence number)",
+			msg:  msgs[2],
+			op: func(hApp *app.HeimdallApp) error {
+				acc := hApp.AccountKeeper.GetAccount(sdkCtx, hmTypes.BytesToHeimdallAddress([]byte(address)))
+				txBroadcaster.lastSeqNo = acc.GetSequence() + 1
+				return nil
+			},
+			expResCode: 4,
+			expErr:     true,
+		},
 	}
 
-	for index, test := range testData { //nolint
-		index := index
-		test := test
-
-		t.Run(strconv.Itoa(index), func(t *testing.T) {
-			t.Parallel()
-
-			// create and send checkpoint message
-			msg := checkpointTypes.NewMsgCheckpointBlock(
-				test.Proposer,
-				test.StartBlock,
-				test.EndBlock,
-				test.RootHash,
-				test.AccountRootHash,
-				test.BorChainID,
-			)
-
-			err := _txBroadcaster.BroadcastToHeimdall(msg, nil)
-			assert.Empty(t, err, "Error broadcasting tx to heimdall", err)
+	for _, tc := range testCases {
+		if tc.expErr {
+			updateMockData(t)
+		}
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.op != nil {
+				err := tc.op(heimdallApp)
+				require.NoError(t, err)
+			}
+			txRes, err := txBroadcaster.BroadcastToHeimdall(tc.msg, nil, testOpts)
+			require.NoError(t, err)
+			require.Equal(t, tc.expResCode, txRes.Code)
+			accSeq, err := heimdallApp.AccountKeeper.GetSequence(sdkCtx, hmTypes.BytesToHeimdallAddress([]byte(address)))
+			require.NoError(t, err)
+			require.Equal(t, txBroadcaster.lastSeqNo, accSeq)
+
+			if tc.tearDown != nil {
+				err := tc.tearDown(heimdallApp)
+				require.NoError(t, err)
+			}
 		})
 	}
 }
+
+func createTestApp(isCheckTx bool, testOpts *helper.TestOpts) (*app.HeimdallApp, sdk.Context, cosmosCtx.CLIContext) {
+	app := app.Setup(isCheckTx, testOpts)
```

Comment:
maybe we want to change this var name, as it collides with the imported package name